stdlib/json.gr

```diff
@@ -0,0 +1,2112 @@
+      {
+        indentation: NoIndentation,
+        arrayFormat: CompactArrayEntries,
+        objectFormat: CompactObjectEntries,
+        lineEnding: NoLineEnding,
+        finishWithNewLine: false,
+        escapeAllControlPoints: false,
+        escapeHTMLUnsafeSequences: false,
+        escapeNonASCII: false,
+      },
+    PrettyAndSafe =>
+      {
+        indentation: IndentWithSpaces(2),
+        arrayFormat: OneArrayEntryPerLine,
+        objectFormat: OneObjectEntryPerLine,
+        lineEnding: LineFeed,
+        finishWithNewLine: true,
+        escapeAllControlPoints: true,
+        escapeHTMLUnsafeSequences: true,
+        escapeNonASCII: true,
+      },
+    CompactAndSafe =>
+      {
+        indentation: NoIndentation,
+        arrayFormat: CompactArrayEntries,
+        objectFormat: CompactObjectEntries,
+        lineEnding: NoLineEnding,
+        finishWithNewLine: false,
+        escapeAllControlPoints: true,
+        escapeHTMLUnsafeSequences: true,
+        escapeNonASCII: true,
+      },
+    Custom{
+      indentation,
+      arrayFormat,
+      objectFormat,
+      lineEnding,
+      finishWithNewLine,
+      escapeAllControlPoints,
+      escapeHTMLUnsafeSequences,
+      escapeNonASCII,
+    } =>
+      {
+        indentation,
+        arrayFormat,
+        objectFormat,
+        lineEnding,
+        finishWithNewLine,
+        escapeAllControlPoints,
+        escapeHTMLUnsafeSequences,
+        escapeNonASCII,
+      },
+  }
+  let writer = makeJsonWriter(format, buf)
+  let error = writer.emit(json)
+
+  match (error) {
+    None => Ok(Buffer.toString(buf)),
+    Some(e) => Err(e),
+  }
+}
+
+/**
+ * Represents errors for JSON parsing along with a human readable text message.
+ */
+provide enum JsonParseError {
+  UnexpectedEndOfInput(String),
+  UnexpectedToken(String),
+  InvalidUTF16SurrogatePair(String),
+}
+
+/**
+ * Internal data structure used during parsing.
+ */
+record JsonParserState {
+  string: String,
+  bufferParse: Buffer.Buffer,
+  mut currentCodePoint: Number,
+  mut pos: Number,
+  mut bytePos: Number,
+}
+
+let isInterTokenWhiteSpace = (codePoint: Number) => {
+  match (codePoint) {
+    0x0009 => true, // tab
+    0x000A => true, // line feed
+    0x000D => true, // carriage return
+    0x0020 => true, // space
+    _ => false,
+  }
+}
+
+let _END_OF_INPUT = -1
+
+@unsafe
+let rec readCodePoint = (bytePosition: Number, string: String) => {
+  from WasmI32 use { (+), (<) }
+
+  let strPtr = WasmI32.fromGrain(string)
+
+  let byteSize = WasmI32.load(strPtr, 4n)
+
+  let bytePositionW32 = coerceNumberToWasmI32(bytePosition)
+
+  let ptr = strPtr + 8n + bytePositionW32
+
+  if (bytePositionW32 < byteSize) {
+    let codePoint = getCodePoint(ptr)
+    tagSimpleNumber(codePoint)
+  } else {
+    _END_OF_INPUT
+  }
+}
+
+let codePointUTF8ByteCount = (usv: Number) => {
+  if (!Char.isValid(usv)) {
+    fail "Impossible: JSON parser encountered an invalid unicode scalar value in codePointUTF8ByteCount"
```

Comment:
Why is this Impossible? It seems like this is used during parsing and a string could contain an invalid usv. @ospencer what do you think here?

Reply:
I'm still unclear on why this is impossible if we have a string from not-grain but I'll resolve 🤷